shell/common/shell.cc

```diff
@@ -418,23 +347,48 @@ const TaskRunners& Shell::GetTaskRunners() const {
   return task_runners_;
 }
 
-fml::WeakPtr<Rasterizer> Shell::GetRasterizer() {
-  FML_DCHECK(is_setup_);
-  return rasterizer_->GetWeakPtr();
+fml::WeakPtr<ShellIOManager> Shell::GetIOManager() const {
+  if (io_manager_) {
+    return io_manager_->GetWeakPtr();
+  }
+  if (io_manager_future_.valid()) {
+    return io_manager_future_.get();
+  }
+  return {};
 }
 
-fml::WeakPtr<Engine> Shell::GetEngine() {
-  FML_DCHECK(is_setup_);
-  return weak_engine_;
+fml::WeakPtr<Rasterizer> Shell::GetRasterizer() const {
+  if (rasterizer_) {
+    return rasterizer_->GetWeakPtr();
+  }
+  if (rasterizer_future_.valid()) {
+    return rasterizer_future_.get();
+  }
+  return {};
 }
 
-fml::WeakPtr<PlatformView> Shell::GetPlatformView() {
-  FML_DCHECK(is_setup_);
-  return platform_view_->GetWeakPtr();
+fml::WeakPtr<Engine> Shell::GetEngine() const {
+  if (engine_) {
+    return engine_->GetWeakPtr();
+  }
+  if (engine_future_.valid()) {
+    return engine_future_.get();
+  }
+  return {};
+}
+
+fml::WeakPtr<PlatformView> Shell::GetPlatformView() const {
+  if (platform_view_) {
+    return platform_view_->GetWeakPtr();
+  }
+  if (platform_view_future_.valid()) {
+    return platform_view_future_.get();
+  }
+  return {};
 }
 
-DartVM* Shell::GetDartVM() {
-  return &vm_;
+fml::WeakPtr<Shell> Shell::GetShell() const {
```

Comment:
WeakPtrs are only safe for access on the thread they were created on.  It looks like there are violations of this in several places in this code.

It also is concerning that this is getting used from the destructor of the shell in post task calls - as in, it seems like there's some attempt to refernece the memory of an object whose destructor has potentially completed.

Reply:
What do yu think of holding Shell with std::shared_ptr by AndroidShellHolder and EmbedderEngine, let Shell extends std::enable_shared_from_this, then we can use std::weak_ptr.lock()  on each closure to access Shell.

Reply:
I think for that scale of change it'd be really nice if we had a design document that outlined the rationale fo rdoing this, the impact of it, and what alternative approaches are being rejected. It would help give the engine team a chance to evaluate what's becoming a fairly large change, and allow for some more comments/refinement than we can likely achieve in a GitHub pull request format.